library/core/src/slice/mod.rs

```diff
@@ -3594,6 +3602,166 @@ impl<T> [T] {
+    /// of the slice is a multiple of the alignment.
+    ///
+    /// # Panics
+    ///
+    /// This method requires the alignment to be a multiple (larger than 1) of the alignment of
+    /// the slice element's alignment.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// unsafe {
+    ///     let bytes: [u8; 7] = [1, 2, 3, 4, 5, 6, 7];
+    ///     let ints = bytes.aligned_subslice(std::mem::align_of::<u32>());
+    ///     assert_eq!(ints.len(), 1);
+    /// }
+    /// ```
+    #[must_use]
+    #[unstable(feature = "slice_align_to_ish", issue = "none")]
+    #[inline]
+    pub fn aligned_subslice(&self, align: usize) -> &[T] {
+        let size = crate::mem::size_of::<T>();
+        assert!(
+            size < align,
+            "aligned_subslice does nothing for alignments below or at the element type's alignment"
+        );
+        assert!(
+            align % size == 0,
+            "aligned_subslice only works for alignments that are multiples of the element's size"
+        );
+        let offset = self.as_ptr().addr() % align;
+        // SAFETY: See the `align_to_mut` method for the detailed safety comment.
+        let end_offset = unsafe { self.as_ptr().offset(self.len() as isize) }.addr() % align;
+        let end = self.len() - (align / size - end_offset);
+        &self[offset..end]
+    }
+
+    /// Get a subslice where the first element is aligned to a given alignment and the length
+    /// of the slice is a multiple of the alignment.
+    ///
+    /// # Panics
+    ///
+    /// This method requires the alignment to be a multiple (larger than 1) of the alignment of
+    /// the slice element's alignment.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// unsafe {
+    ///     let mut bytes: [u8; 7] = [1, 2, 3, 4, 5, 6, 7];
+    ///     let ints = bytes.aligned_subslice_mut(std::mem::align_of::<u32>());
+    ///     assert_eq!(ints.len(), 1);
+    /// }
+    /// ```
+    #[must_use]
+    #[unstable(feature = "slice_align_to_ish", issue = "none")]
+    #[inline]
+    pub fn aligned_subslice_mut(&mut self, align: usize) -> &mut [T] {
+        let size = crate::mem::size_of::<T>();
+        assert!(
+            size < align,
+            "aligned_subslice does nothing for alignments below or at the element type's alignment"
+        );
+        assert!(
+            align % size == 0,
+            "aligned_subslice only works for alignments that are multiples of the element's size"
+        );
+        let offset = self.as_ptr().addr() % align;
+        // SAFETY: See the `align_to_mut` method for the detailed safety comment.
+        let end_offset = unsafe { self.as_ptr().offset(self.len() as isize) }.addr() % align;
+        let end = self.len() - (align / size - end_offset);
+        &mut self[offset..end]
+    }
+
+    /// Transmute the slice elements to another (smaller) type, giving a longer slice.
```

Comment:
Is there a reason to identify it as *smaller*? This seems useful for 1:1 mapping, too, e.g. `u8` <> `i8`.